### PR TITLE
Refactor(token-store-internal) - stores segregation

### PR DESCRIPTION
**Pull request by joan41868**
**Related issue(s)**:
Not closing any issues yet.

**Summary of the change**:
Implementation of basic store responsibility segregation.
Implementation of OwnerIdentifier object, required for InvertibleHashMap.

**External impacts**:
None.

**Applicable documentation**
- [x] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java [hunk 1]
```diff
@@ -116,6 +103,7 @@
 		static final int NUM_0120_CHILDREN = 10;
 		static final int NUM_0130_CHILDREN = 10;
 		static final int NUM_0140_CHILDREN = 10;
+		static final int NFTOKENS = 11;
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/ServicesState.java:106]: I think that you need initialize the nft tokens in `initialize(MerkleInternal previous)` similar to the previous (e.g. `scheduledTxs()` or `runningHashLeaf()`. Check the whole file where `scheduledTxs` has been added and similar to that add nfts.

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 2]
```diff
@@ -1972,6 +1832,10 @@ public ExchangeRates midnightRates() {
 		return state.tokens();
 	}
 
+	public FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier> nfTokens(){
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:1835]: nit: space before `{`

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleUniqueToken.java [hunk 3]
```diff
@@ -119,6 +123,7 @@ public void serialize(SerializableDataOutputStream out) throws IOException {
 	}
 
 	/* --- FastCopyable --- */
+
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleUniqueToken.java:126]: nit: unnecessary line

File: hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java [hunk 4]
```diff
@@ -199,7 +166,7 @@ public ResponseCodeEnum dissociate(AccountID aId, List<TokenID> targetTokens) {
 					return TOKEN_NOT_ASSOCIATED_TO_ACCOUNT;
 				}
 				if (!tokens.get().containsKey(fromTokenId(tId))) {
-					/* Expired tokens that have been removed from state (either because they
+					/* Expired tokens that have been reBaseTokenStoremoved from state (either because they
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java:169]: nit: revert this change I guess

File: hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java [hunk 5]
```diff
@@ -221,8 +188,9 @@ public ResponseCodeEnum dissociate(AccountID aId, List<TokenID> targetTokens) {
 					if (!isTokenDeleted && !isTokenExpired) {
 						return TRANSACTION_REQUIRES_ZERO_TOKEN_BALANCES;
 					}
+					// TODO Works for common, but for UN should be a different type of transfer
 					if (!isTokenDeleted) {
-						/* Must be expired; return balance to treasury account. */
+						/* Must be expired; return balaBaseTokenStorence to treasury account. */
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java:193]: nit: revert this change

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package com.hedera.services.store.tokens.common;/*
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java:1]: nit: license should begin on line `3`

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonTokenStore.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
+package com.hedera.services.store.tokens.common;/*
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonTokenStore.java:1]: nit: license should begin on line 3

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/OwnerIdentifier.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
+package com.hedera.services.store.tokens.unique;/*
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/OwnerIdentifier.java:1]: nit: license should begin on line 3

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonTokenStore.java (hunk 7 above)
Comment (failfmi) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonTokenStore.java:1]: I am not sure from where this is copied, should be 2018 - 2021

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/OwnerIdentifier.java (hunk 8 above)
Comment (failfmi) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/unique/OwnerIdentifier.java:1]: 2018 - 2021

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueStore.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package com.hedera.services.store.tokens.unique;/*
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueStore.java:1]: nit: license should begin on line 3
2018 - 2021

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
+package com.hedera.services.store.tokens.unique;
+/*
```
Comment (failfmi) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java:2]: nit: license should begin on line 3

2018-2021

**Review by failfmi — COMMENTED** [review]
It would be best if we revert the changes with the imports.

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java (hunk 6 above)
Comment (failfmi) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java:1]: 2018-2021

File: hedera-node/src/test/java/com/hedera/services/context/ServicesContextTest.java [hunk 11]
```diff
@@ -159,18 +125,8 @@
...
-import static org.junit.jupiter.api.Assertions.assertSame;
-import static org.mockito.BDDMockito.any;
-import static org.mockito.BDDMockito.given;
-import static org.mockito.BDDMockito.inOrder;
-import static org.mockito.BDDMockito.mock;
-import static org.mockito.BDDMockito.spy;
-import static org.mockito.BDDMockito.verify;
-import static org.mockito.BDDMockito.when;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.BDDMockito.*;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/context/ServicesContextTest.java:129]: We should have the same InteliJ config for imports

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java [hunk 12]
```diff
@@ -116,6 +119,7 @@
...
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/ServicesState.java:122]: I think that a better term would be `UNIQUE_TOKENS`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 13]
```diff
@@ -441,6 +448,7 @@
 	private MetaAnswers metaAnswers;
 	private RecordCache recordCache;
 	private TokenStore tokenStore;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:450]: Lets use `CommonTokenStore`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 14]
```diff
@@ -441,6 +448,7 @@
...
+	private UniqueStore uniqueStore;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:451]: nit: Maybe we should rename it to `UniqueTokenStore` for consistency?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java [hunk 15]
```diff
@@ -96,13 +96,21 @@
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_WAS_DELETED;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TRANSACTION_REQUIRES_ZERO_TOKEN_BALANCES;
 
+// Common Fungible -> what we have now
+// Unique Non-fungible -> what we wanted to do with NFT
+// Unique Fungible -> not discussed
+
+// CF -> HTS, supply, amount
+// UF -> HTS
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java:104]: nit: remove the comments

File: hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java [hunk 16]
```diff
@@ -96,13 +96,21 @@
...
+
 /**
  * Provides a managing store for arbitrary tokens.
  *
  * @author Michael Tinker
  */
-public class HederaTokenStore extends HederaStore implements TokenStore {
-	private static final Logger log = LogManager.getLogger(HederaTokenStore.class);
+// TODO make it abstract and `BaseTokenStore`. Use CommonTokenStore everywhere for now
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java:111]: nit: remove the comments

File: hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java [hunk 17]
```diff
@@ -381,6 +397,7 @@ private ResponseCodeEnum changeSupply(
 		});
 	}
 
+	// Most probably will go directly to base
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/BaseTokenStore.java:400]: nit: remove comment

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.hedera.services.store.tokens.TokenStore;
+
+public interface CommonStore extends TokenStore {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java:25]: I think that we can move the following methods to `CommonStore`:
```Java
ResponseCodeEnum burn(TokenID tId, long amount);
	ResponseCodeEnum mint(TokenID tId, long amount);
	ResponseCodeEnum wipe(AccountID aId, TokenID tId, long wipingAmount, boolean skipKeyCheck);
	ResponseCodeEnum freeze(AccountID aId, TokenID tId);
	ResponseCodeEnum update(TokenUpdateTransactionBody changes, long now);
	ResponseCodeEnum unfreeze(AccountID aId, TokenID tId);
	ResponseCodeEnum grantKyc(AccountID aId, TokenID tId);
	ResponseCodeEnum revokeKyc(AccountID aId, TokenID tId);
	ResponseCodeEnum associate(AccountID aId, List<TokenID> tokens);
	ResponseCodeEnum dissociate(AccountID aId, List<TokenID> tokens);
	ResponseCodeEnum adjustBalance(AccountID aId, TokenID tId, long adjustment);
```

This will force us to implement all those methods for `UniqueStore` and figure out whether we need to have different method signatures / overload them

[the remark this answers is not in the document]
Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java:25]: Renaming it to CommonTokenStore would force me to rename the class implementation of the interface as well.
I can go for CommonTokenStore - interface, and CommonTokenStoreImpl - implementation too.
Another variant would be ICommonTokenStore

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+			if (!merkleToken.hasSupplyKey()) {
+				return TOKEN_HAS_NO_SUPPLY_KEY;
+			}
+			var mintResult = super.mint(tId, 1);
+			final var suppliedTokens = uniqueTokensSupply.get();
+			final var eId = EntityId.fromGrpcTokenId(tId);
+			final var owner = get(tId).treasury(); // get next available serial num here as well
+			final int serialNum = 0;
+			// TODO serialNum
+			// When a merkleUniqueToken is created, the next present serial number must be available as a method
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java:85]: This branch should be based on the `TokenCreate` changes branch. We should add `currentSerialNumber` property in the `MerkleToken` entity in order to implement this logic

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+		}
+		if (start < 0){
+			throw new IllegalArgumentException("Start index cannot be negative");
+		}
+
+		return uniqueTokensSupply.get().inverseGet(token, start);
+	}
+
+	@Override
+	public Iterator<MerkleUniqueTokenId> getByTokenFromIdxToIdx(final MerkleUniqueToken token, final int start, final int end) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java:117]: Not sure that this would work out. We should discuss it

File: hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+		if(end < start)
+			throw new IllegalArgumentException("Start index cannot be bigger than end. "+ start + " > "+ end);
+
+		if (end > uniqueTokensSupply.get().size())
+			throw new IllegalArgumentException("End index "+ end + " is larger than size.");
+
+		final var eId = EntityId.fromGrpcAccountId(aId);
+		var tokenSupply = uniqueTokensSupply.get();
+		final var ownerIdentifier = new OwnerIdentifier(eId);
+		return tokenSupply
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java:152]: `inverseGet` does the same logic more effectively. We should be using that functionality.

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenCreateTransitionLogic.java [hunk 22]
```diff
@@ -57,7 +55,8 @@
 	private final Function<TransactionBody, ResponseCodeEnum> SEMANTIC_CHECK = this::validate;
 
 	private final OptionValidator validator;
-	private final TokenStore store;
+	private final TokenStore store; // CommonTokenStore
+	// UniqueTokenStore
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenCreateTransitionLogic.java:59]: nit: remove comment

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java [hunk 23]
```diff
@@ -48,29 +46,44 @@
...
 			TransactionContext txnCtx
 	) {
-		this.store = store;
+		this.commonStore = commonStore;
+		this.uniqueStore = uniqueStore;
 		this.txnCtx = txnCtx;
 	}
 
+	// TODO temporary fix
+	public TokenMintTransitionLogic(
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java:65]: Q: Why don't we use the other constructor?

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (Daniel-K-Ivanov) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/common/CommonStore.java:35]: got it

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java (hunk 23 above)
Comment (joan41868) [reply hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java:65]: Because the current InvertibleHashMap implementation is not a descendant of MerkleNode, and thus cannot  be loaded in ServicesContext
